### PR TITLE
feat: Bb 276 implement calculate util

### PR DESCRIPTION
**Pull request by GDamaso**
# Description

Created an util function to calculate NPK agronomic balances and crop removal balances. The function should be modular but is being built at first focusing on the raspberry calculation spreadsheet.

## Type of changes

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have already been accepted and merged

### BB Field for testing calculation
BB Field for testing calculation

### BB Crop info
BB Crop info

### Spreadsheet calculation test data
Crop Type |   | Crop |   | Yield (tons/ac) |   |  
-- | -- | -- | -- | -- | -- | --
Berry |   | Raspberry |   | 5 |   |  
  |   |   |   |   |   |   |   |   |  
  |   |   |   |   |   |   |   |   |  
Will plants be pruned? |   | Where will prunings go? |   | Will sawdust be applied prior to fertilization? |   |   |  
Yes |   | Left in row |   | Yes

### Soil test data
NO3-N (ppm), nitrate-nitrogen |   |   |   | 3.4
-- | -- | -- | -- | --
  |   |   |   |  
P (ppm), phosphorus |   |   |   | 20
  |   |   |   |  
K (ppm), potassium |   |   |   | 130
  |   |   |   |  
pH |   |   |   | 6.2

### Leaf test data
Leaf tissue P (%) |   | Leaf tissue K (%)
-- | -- | --
0.16-0.18 |   | 1.00-1.25

### Calculation results vs BB console log results

image

image

---

Thanks for the PR!

Deployments, as required, will be available below:
- [Frontend](https://nr-sustainment-capstone-2024-287-frontend.apps.silver.devops.gov.bc.ca)
- [Backend](https://nr-sustainment-capstone-2024-287-backend.apps.silver.devops.gov.bc.ca/api/health)

Please create PRs in draft mode.  Mark as ready to enable:
- [Analysis Workflow](https://github.com/bcgov/nr-sustainment-capstone-2024/actions/workflows/analysis.yml)

After merge, new images are deployed in:
- [Merge Workflow](https://github.com/bcgov/nr-sustainment-capstone-2024/actions/workflows/merge.yml)

### CODE REVIEW

File: frontend/src/Commons/Forms/InputModules/FieldsAndSoil/FieldsAndSoil.tsx [hunk 1]
```diff
@@ -109,6 +110,17 @@ const FieldsAndSoilComponent: FC<InputModuleProps> = ({
   const addFieldData = (values: FieldDetailInterface): void => {
     setTimeout(() => {
       const farmInfo: FarmDetailsInterface = { ...farmDetails };
+      const noSoilTestVal = {
+        ...values.SoilTest,
+        ValP: Infinity,
```
Comment (kcaparas1630) [inline frontend/src/Commons/Forms/InputModules/FieldsAndSoil/FieldsAndSoil.tsx:115]: just a random question, what does Infinity do?

File: frontend/src/Utils/Calculate/Calculate.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+// This function will be updated in the next ticket
+// It is mostly a proof of correct calculations being applied
+// For this reason, it is scoped to the first field first crop only at this point
+function Calculate(field: FieldDetailInterface) {
+  const crop = field.Crops[0];
+  const calcTable: CalculationTable = calcData;
+  const agronomicBalance: AgronomicBalanceInterface = { N: 0, P: 0, K: 0 };
+  const cropRemovalBalance: CropRemovalBalanceInterface = { P: 0, K: 0 };
+  const isRemovedFromField = field.Crops[0].whereWillPruningsGo === 'Removed from field';
+  console.log(field.Crops[0].whereWillPruningsGo);
```
Comment (kcaparas1630) [inline frontend/src/Utils/Calculate/Calculate.ts:88]: leftover log?

**Review by kcaparas1630 — APPROVED** [review]
I'm not blocking you on this. I don't see anything that is wrong to me. Just some minor questions that I need to be answered for my own curiosity.

I've tested it and it LGTM! 🚀

File: frontend/src/Commons/Forms/InputModules/FieldsAndSoil/FieldsAndSoil.tsx (hunk 1 above)
Comment (GDamaso) [reply frontend/src/Commons/Forms/InputModules/FieldsAndSoil/FieldsAndSoil.tsx:115]: We use ranges for the calculation, which means that infinity here will always be greater than the min range it's being checkes for. It's used when the user does not provide soil/leaf tests

File: frontend/src/Utils/Calculate/Calculate.ts (hunk 2 above)
Comment (GDamaso) [reply frontend/src/Utils/Calculate/Calculate.ts:88]: Good catch!

**Pull request merged**